### PR TITLE
Processing data in models

### PR DESCRIPTION
**Pull request by danielochuba**
# Blog app - Processing data in models
### In this pull request I did the following: 
- Created models
    - `User`
    - `Post`
    - `Comment`
    - `Like`
- Established relationships 
    - a  `User` has many `Post`
    - a `User` has many `Comments`
    - a `Post` has many `Likes`
    - a `Post` has many `Comments`
    - a `Post` belongs to a `User`
    - Many `Comment` belongs to a `User`
    - Many `Comment` belongs to a `Post`
- Create custom methods
     - Users
              - A method that returns the 3 most recent posts for a given user.
     - Posts
             - A method that updates the posts counter for a user.
             - A method which returns the 5 most recent comments for a given post.
     - Comments
          - A method that updates the comments counter for a post.
     - Likes
          - A method that updates the likes counter for a post.

### CODE REVIEW

File: app/models/comment.rb [hunk 1]
```diff
@@ -0,0 +1,4 @@
+class Comment < ApplicationRecord
+  belongs_to :author, class_name: 'User', foreign_key: 'user_id'
```
Comment (elisha2kyakpo1) [inline app/models/comment.rb]: - [ ] The user foreign key in the comments table is user_id, this is also the same with the ERD. Kindly address this code to reflect what is in the schema.
- [ ] Kindly add a function that will increase the counter of the comments when a user adds a comment to a post.

foreign-key

File: app/models/post.rb [hunk 2]
```diff
@@ -0,0 +1,19 @@
+class Post < ApplicationRecord
+  belongs_to :author, class_name: 'User', foreign_key: 'author_id'
+  has_many :comments
+  has_many :likes
+
+  def update_user_posts_counter
+    author.update(post_count: author.authored_posts.count)
+  end
+
```
Comment (elisha2kyakpo1) [inline app/models/post.rb]: - [ ] The purpose of these counter functions is to increase the counter every time a post is saved in the database. Kindly make use of the model callback functions such as `after_save`. This function will save a counter every time a post is saved.

counter


```ruby
  after_save :update_user_posts_counter
  private

  def update_user_posts_counter
    author.increment!(:post_counter)
  end
```

File: app/models/post.rb [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+  # Method to return the 5 most recent comments for a given post
+  def recent_comments(limit = 5)
+    comments.order(created_at: :desc).limit(limit)
+  end
+
+  # Method to update the likes counter for a post
+  def update_likes_counter
+    update(likes_counter: likes.count)
+  end
```
Comment (elisha2kyakpo1) [inline app/models/post.rb]: - [ ] The purpose of these counter functions is to increase the counter every time a like is saved in the database. Kindly make use of the model callback functions such as `after_save`. This function will save a counter every time a like is saved. 
__Kindly remove this function from here and take it into the likes model where it should belong__.

```ruby
  after_save :update_likes_counter
  private

  def update_llikes_counter
    author.increment!(:likes_counter)

**Review by elisha2kyakpo1 — CHANGES_REQUESTED** [review]
Hi 👋 @danielochuba, 

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!

## Highlights

✔️ The PR title and description are good and professional.
✔️ No linter errors.
✔️ Descriptive readme file.
✔️ Associations nicely done between a post and a user (author id relationship)👏

## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

Cheers, and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Comment by danielochuba** [comment]
@elisha2kyakpo1 Thank you for that wonderful feedback, I appreciate

File: app/models/comment.rb [hunk 4]
```diff
@@ -1,4 +1,11 @@
 class Comment < ApplicationRecord
   belongs_to :author, class_name: 'User', foreign_key: 'user_id'
```
Comment (elisha2kyakpo1) [inline app/models/comment.rb]: - [ ] This issue here has not been addressed, your comments table (in the schema) knows a user as a user, with an ID of `user_id` and not `author_id`. The only table to carry the name `author_id` as the foreign key is the post table. You can change the `belongs_to :author` line to `belongs_to :user`.

File: app/models/post.rb [hunk 5]
```diff
@@ -3,17 +3,26 @@ class Post < ApplicationRecord
...
+  after_save :update_likes_counter
+
+  private
+  def update_user_posts_counter
+    author.update(post_count: author.authored_posts.count)
+  end
+
   def update_likes_counter
     update(likes_counter: likes.count)
   end
```
Comment (elisha2kyakpo1) [inline app/models/post.rb]: - [ ] You already have this method in the likes model, we don't want this function here as earlier reviewed.

**Review by elisha2kyakpo1 — CHANGES_REQUESTED** [review]
Hi 👋 @danielochuba, 

Great work on making the changes requested by a previous reviewer 👏 
You've done well implementing some of the requested changes, but there are still some which aren't addressed yet.


 ## Highlights

👍 You did well fixing almost all the issues raised in my earlier review.

 ## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._


Cheers, and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**


------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: app/models/comment.rb (hunk 4 above)
Comment (danielochuba) [reply app/models/comment.rb]: comment

@elisha2kyakpo1 what do you have to say about this

I invited you on a zoom call for further explanations

**Review by IngDavidTM — APPROVED** [review]
Hi @danielochuba 🙋‍♂️, 

# ✔️   Status: APPROVED 🏆 🎉 🎊 

You have done a great job on this project, keep it up 👍 🥇. Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations!  🎉💯🌟 

## Top highlight 🎯

- The Pull Request follows the GitFlow ✅ 
- The PR title and description are professional and descriptive ✅ 
- Linters are working properly ✅ 


## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**